scripts/document_indexer.py

```diff
@@ -0,0 +1,315 @@
+
+
+def chunk_document(
+    doc_path: str,
+    chunk_size: int,
+    chunk_step: int,
+) -> Tuple[int, List[str], List[Dict[str, str]], Dict[str, int]]:
+    """
+    Split documents into chunks
+    :param doc_path: the path of the documents
+    :param chunk_size: the size of the chunk
+    :param chunk_step: the step size of the chunk
+    """
+    texts = []
+    metadata_list = []
+    file_count = 0
+    chunk_id_to_index = dict()
+
+    enc = tiktoken.encoding_for_model("gpt-3.5-turbo")
+
+    # traverse all files under dir
+    print("Split documents into chunks...")
+    for root, dirs, files in os.walk(doc_path):
+        for name in files:
+            f = os.path.join(root, name)
+            print(f"Reading {f}")
+            try:
+                title, content, _, _ = text_parser(f)
+                file_count += 1
+                if file_count % 100 == 0:
+                    print(f"{file_count} files read.")
+
+                if len(content) == 0:
+                    continue
+
+                chunks = chunk_str_overlap(
+                    content.strip(),
+                    num_tokens=chunk_size,
+                    step_tokens=chunk_step,
+                    separator="\n",
+                    encoding=enc,
+                )
+                source = os.path.sep.join(f.split(os.path.sep)[4:])
+                for i in range(len(chunks)):
+                    # custom metadata if needed
+                    metadata = {
+                        "source": source,
+                        "title": title,
+                        "chunk_id": i,
+                    }
+                    chunk_id_to_index[f"{source}_{i}"] = len(texts) + i
+                    metadata_list.append(metadata)
+                texts.extend(chunks)
+            except Exception as e:
+                print(f"Error encountered when reading {f}: {traceback.format_exc()} {e}")
+    return file_count, texts, metadata_list, chunk_id_to_index
+
+
+if __name__ == "__main__":
+    # parse arguments
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-d",
+        "--doc_path",
+        help="the path of the documents",
+        type=str,
+        default="",
+    )
+    parser.add_argument(
+        "-c",
+        "--chunk_size",
+        help="the size of the chunk",
+        type=int,
+        default=64,
+    )
+    parser.add_argument(
+        "-s",
+        "--chunk_step",
+        help="the step size of the chunk",
+        type=int,
+        default=64,
+    )
+    parser.add_argument(
+        "-o",
+        "--output_path",
+        help="the path of the output",
+        type=str,
+        default="",
+    )
+    args = parser.parse_args()
+
+    file_count, texts, metadata_list, chunk_id_to_index = chunk_document(
+        doc_path=args.doc_path,
+        chunk_size=args.chunk_size,
+        chunk_step=args.chunk_step,
+    )
+    embeddings = HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2")
```

Comment:
is it possible to use llm module from taskweaver so that all the embedding configuration would be reused

Reply:
in theory yes, but I also used langchain that can only recognize its own embedding class. this means that we need to implement all the other logic